service/service.hpp

```diff
@@ -185,21 +188,44 @@ namespace raven
         DLOG_IF_F(INFO, cfg.config_read_only_key.has_value(), "cfg.config_read_only_key: %s",
                   cfg.config_read_only_key.value().value().c_str());
 
-        auto answer = db_.config_load(cfg);
-        if (answer.request_state == convert_request_state.at(request_state::success)) {
-            config_clients_registry_.at(sock.fileno()) += answer.config_id;
-            answer.config_id = config_clients_registry_.at(sock.fileno()).get_last_id();
+        config_id_st id;
+        if (cfg.config_key) {
+            id = db_.get_config_id(cfg.config_key.value());
+        } else if (cfg.config_read_only_key) {
+            id = db_.get_config_id(cfg.config_read_only_key.value());
+        } else {
+            send_answer(sock, request_state::unknown_request);
+            return ;
         }
-        prepare_answer(sock, answer);
+        if (db_.fail()) {
+            if (db_.get_state() == db_state::unknow_config_key) {
+                send_answer(sock, request_state::unknown_key);
+                return ;
+            }
+            send_answer(sock, request_state::db_error);
+            return ;
+        }
+
+        std::string name = db_.get_config_name(id);
+        if (db_.fail()) {
+            send_answer(sock, request_state::db_error);
+            return ;
+        }
+
+        config_id_st temp_id = config_clients_registry_.at(sock.fileno()).insert_db_id(id);
+        // TODO get permision for the key and store it with the id
+        // TODO add ref_counting
+        // TODO get full config in cache
+        send_answer(sock, config_load_answer{name, temp_id, convert_request_state.at(request_state::success)});
```

Comment:
Unless send_answer is overloaded, it seems the last thing that's done in all cases is call "send_answer". I think using a local variable to store the answer could make the code a lot more readable. If you feel you have to many nested ifs, you could use a `done` label combined with `goto` (le gasp, he suggested it) to avoid some of the nesting.

Reply:
send_answer is overloaded, not all answer share the same type.
Also not all operation end up with a send_answer, most of them do because each request require an answer, but in the case of subscribe and unsubscribe, they fire subscribtion event after sending back the answer, and it need to be in that particular order

Reply:
Actually, I think this could be a use-case for exceptions. A priori all failures are handled in a similar way. You could add a function "throw_failure" in the db class which would throw the proper error if fail() returns true. For the middle if (the one which changes the error state) you'd have to catch and rethrow. This mechanism doesn't need to be implemented in the db directly, so you could keep it all external if you want.

Reply:
I'm not too confortable to use exception here either, some db_state may not be fatal and the operation requested may still be continuable, for now It's only possible to get the last state of an operation (mostly because i didn't thought the error and exception handling completly through yet), but in the future my idea is that the `get_state` returns a bitfield that hold much more informations that just a single state.
If we use exception, we might interupt the operation when it could have gone through, reporting afterward with the bitfield what happen and let the user judge if the result is satisfactory or not

We can discuss it more in a new issue that address that particular topic ?

Reply:
Ok, I'll talk to you if we ever get the chance. I still haven't reviewed the whole PR (I will probably skip the tests) so I can't approve this yet.